contributor-docs/adrs/adr-003-prop-norms.md

```diff
@@ -0,0 +1,70 @@
+# ADR 003: Prop norms in Primer React Components
+
+## Status
+
+Proposed
+
+## Context
+
+Today we have:
+
+- Implicit conventions not documented anywhere but consistently reflected in our code (e.g., the type of the `sx` prop)
+- Explicit plans to change some of those (e.g., the deprecation of styled-system props)
+- Inconsistencies in our implementation (e.g., when components accept a `ref` prop)
+
+This ADR aims to unify conversations about prop APIs, codify our decisions, and sequence the work to get there.
+
+## Decision
+
+### 🟢 `sx`
+
+All components that ultimately render to the DOM should accept an `sx` prop.
+
+The `sx` prop (of type `SystemStyleObject`) should generally set styles for the root HTML element rendered by the component. An exception would be components like `Dialog`, whose outermost HTML element is a backdrop. In that case, it would be appropriate for `sx` styles to apply to child of the backdrop that is more likely to need styling overrides.
+
+It may make sense to have an additional prop to style a descendant element (e.g., `Dialog` could accept a `backdropSx` prop).
+
+### 🟢 `ref`
+
+All components that ultimately render to the DOM should accept a `ref` prop. That `ref` prop should be passed to the root HTMLElement rendered by the component.
+
+### 🟡 `as`
+
+Only components with a clear need for polymorphism should accept an `as` prop. Reasonable cases include:
+
+- Components that need functionality from the `as` prop, like a button
+- Components whose accessibility are improved by using semantically appropriate HTML elements
+
+When passing an `as` prop to a component, it should be done in a way that is consistent with the component’s intended use.
+
+### 🟡 DOM props: Limited
+
+All components that accept an `as` prop should accept the props for the element specified by the as prop. _Additionally_, some other elements (e.g., `<TextInput>`) should accept the props for their root HTML element when those props are fundamental to the component’s function.
```

Comment:
How do we feel about components accepting some standard DOM props like `id`, `className`, `data-*`? Are these props that all components that render to the DOM should accept? I know having `id` and `data-*` can be helpful for testing purposes.

Reply:
[Strong opinion, weakly held](https://www.saffo.com/02008/07/26/strong-opinions-weakly-held/): It is fine to add support for props that are one-by-one relayed to the DOM (i.e., not en masse via a spread) as they are needed.

In your example, if there was a need for a test attribute, I'd rather accept a prop like `data-test-id` than `id` so the naming reinforces the semantics of the field.